Sources/Virtual Currencies/VirtualCurrencyManager.swift

```diff
@@ -0,0 +1,119 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  VirtualCurrencyManager.swift
+//
+//  Created by Will Taylor on 5/28/25.
+
+import Foundation
+
+protocol VirtualCurrencyManagerType {
+    func virtualCurrencies() async throws -> VirtualCurrencies
+
+    func invalidateVirtualCurrenciesCache()
+}
+
+class VirtualCurrencyManager: VirtualCurrencyManagerType {
+
+    private let identityManager: IdentityManager
+    private let deviceCache: DeviceCache
+    private let backend: Backend
+    private let systemInfo: SystemInfo
+
+    init(
+        identityManager: IdentityManager,
+        deviceCache: DeviceCache,
+        backend: Backend,
+        systemInfo: SystemInfo
+    ) {
+        self.identityManager = identityManager
+        self.deviceCache = deviceCache
+        self.backend = backend
+        self.systemInfo = systemInfo
+    }
+
+    func virtualCurrencies() async throws -> VirtualCurrencies {
+        let appUserID = identityManager.currentAppUserID
+        let isAppBackgrounded = systemInfo.isAppBackgroundedState
+
+        if let cachedVirtualCurrencies = await fetchCachedVirtualCurrencies(
+            appUserID: appUserID,
+            isAppBackgrounded: isAppBackgrounded
+        ) {
+            return cachedVirtualCurrencies
+        }
+
+        let virtualCurrencies = try await fetchVirtualCurrenciesFromBackend(
+            appUserID: appUserID,
+            isAppBackgrounded: isAppBackgrounded
+        )
+
+        cacheVirtualCurrencies(virtualCurrencies, appUserID: appUserID)
+
+        return virtualCurrencies
+    }
+
+    func invalidateVirtualCurrenciesCache() {
+        let appUserID = identityManager.currentAppUserID
+        deviceCache.clearVirtualCurrenciesCache(appUserID: appUserID)
+    }
+
+    private func cacheVirtualCurrencies(
+        _ virtualCurrencies: VirtualCurrencies,
+        appUserID: String
+    ) {
+        guard let virtualCurrenciesData = try? JSONEncoder().encode(virtualCurrencies) else {
+            return
+        }
```

Comment:
Question: should we clear the cache at this point?
I guess this should never happen, but still, if we've just received a new virtual currency object, it means that the one currently in cache is not needed anymore.
```suggestion
        guard let virtualCurrenciesData = try? JSONEncoder().encode(virtualCurrencies) else {
            self.deviceCache.clearVirtualCurrenciesCache(appUserID: appUserID)
            return
        }
```